### ISSUE TITLE
Fully-standalone version

### ISSUE THREAD

**Issue by brunobord** [vole/vole#54]
Vole depends on... itself. If you don't want to use the source and run it via go, that's fine.
But if you want to keep up to date, you _have_ to use BTSync, right? This is far from having it completely standalone. With this setup, it can't really be standalone. i.e. what if I want to put everything (profile and binaries) on a USB stick? (of course, it means I must have access to win/linux/osx binaries, but that's not impossible, technically)

I wonder if it was possible to have a tiny BTSync-like small distribution, we could launch along with the vole web server.

Or just a binary that could run both bittorrent thingy _and_ the web application.

What do you think?

**Comment by marktheunissen** [comment]
That would be great. The issue that re-writing the BitTorrent Sync application is non-trivial, and the protocol is not open. This is still an ongoing question for sure.

**Comment by Fil** [comment]
Are you _obliged_ to use BTSync — it seems to me that it's only one of many possible conduits for our data sync needs. git, rsync, dropbox-like tools (open source preferably) can also do the trick

**Comment by marktheunissen** [comment]
Yes, any tool would work. We have no obligation to use BTSync, however it is just the most effective option right now.

Dropbox-like tools also work, but can you find any that allow you to share a _read-only_ folder to sync? I looked for one, but it seems that none of the major sync applications allowed read-only sync.

Rsync / git would work too, but torrent is a far more efficient protocol and has "solved" the issues of NAT & Firewalls already.

Another option is WebRTC.

**Comment by almet** [comment]
+1 for a WebRTC backend, which would make Vole only dependent on itself.

**Comment by marktheunissen** [comment]
WebRTC is something we should use in some form (e.g. private messages between people where one user doesn't follow another), but it has drawbacks when used as the main syncronization backend:
- It's a new technology only supported in some browsers.
- No UDP packets - so peer discovery on local network may not be ideal
- Web browser's LocalStorage is limited and inefficient. How much data can we store in it? 
- How many TCP connections can be made from a browser? Will having 200+ open connections work?
- How will I export my data for use from other locations? 
- JavaScript in general is inefficient but specifically encryption - limiting what we can do in this area.
- Mobile phone browsers - right now you could use a 'hosted' Vole and access it from your phone - this won't work if we rely on the phone's browser to store the data.
- Where will the JavaScript be served from? Will people be required to download a Vole server even if all it does is deliver static HTML / JS? If we put the application on a server, then we go back to having a central point of failure again, although perhaps not so bad as Twitter / etc. What about security breach of this server?

Advantages:
- The JavaScript sandbox is pretty secure.

[Open Peer](http://openpeer.org/) is doing some work in the arena, but it's still early days.

I'd like to hear more opinions on this, see if anyone can correct my points above.

**Comment by marktheunissen** [comment]
- Leaving the browser open - you need to keep it open to maintain connections, whereas Vole can run as a background service. I feel this is a better user experience, as people otherwise will close their browser and thus nobody will be able to connect to them.

**Comment by bencurthoys** [comment]
I was having a conversation with some friends (before having heard of Vole) about this subject a few weeks ago, and -
I know it's a bit of a blast from the past - if you were looking for a protocol for transporting news articles between servers, why not use NNTP? If each "user account" were a separate newsgroup and everyone ran their own news server, you'd end up with about the same topology. You'd get loads of established libraries and readers for free that way.

**Comment by marktheunissen** [comment]
@bencurthoys That is a blast from the past, and an interesting thought experiment. Would each user end up with all the posts from everyone? How could I just 'follow' individual users?

**Comment by bencurthoys** [comment]
I was thinking each user would have their own newsgroups, on their own server. You'd follow someone by subscribing to their newsgroup. Your reader would want to display an aggregated view, as well as a view split by newsgroup (your personal feed vs looking at someone else's wall).

It was a long time ago that I ran a news server, but... I worked for a software company and we hosted newsgroups as an early form of support forum. They were were outside the big 8 (comp._, news._, etc)  hierarchy that were distributed and mirrored by ISPs, which is "Usenet", they just used the same protocol. I was a heavy user of the Microsoft newsgroups at the time and Microsoft only killed off theirs in 2010: http://arstechnica.com/information-technology/2010/05/microsoft-killing-off-support-newsgroups/ 

Our newsgroups were restricted by account/password, so some were public, others were for paying customers only.

So imagine - if everyone ran their own news server, and created their own newsgroup(s) on it with varying levels of access, public, friends, circles, etc,, then to "follow" someone you would configure your news server to grab the posts from theirs, and subscribe to those groups in your newsreader to read them. 

And as far as I can tell, that's the same topology as Vole, it's just you're (currently) using BTSync and JSON files to manage the syncronisation, Creating a profile, adding it to BTSync, sharing the key, and having someone else use the key to sync your data to their server and then reading it using their Vole server is functionally the same as creating a newsgroup on your NNTP server, someone else setting up their server to sync that group, and then they read the messages that are posted.

I never really looked at NNTP at the protocol level - I just set the server up and got back to the other work - but I think it was designed to solve essentially the same problem and is capable of all the necessary operations. It's interesting how the early web protocols were based on the assumption that everyone would have their own server and everything would be decentralised, which we're more or less coming full circle.

Obviosuly, BTSync is brilliant, and once they add an API so you can automate the creation of profiles and following people a lot of the friction will go away, and I for one am happy to see all the substantial non-infringing uses that are popping up for it.

**Comment by marktheunissen** [comment]
That's something to think about for sure. Thanks for the input.
